osquery/filesystem/linux/proc.cpp

```diff
@@ -118,71 +118,211 @@ unsigned short procDecodePortFromHex(const std::string& encoded_port) {
+    if (fields.size() < 7) {
+      VLOG(1) << "Invalid UNIX socket descriptor found: '" << line
+              << "'. Skipping this entry";
+      continue;
+    }
+
+    SocketInfo socket_info = {};
+    socket_info.socket = fields[6];
+    socket_info.family = AF_UNIX;
+    socket_info.protocol = std::atoll(fields[2].data());
+    socket_info.unix_socket_path = (fields.size() >= 8) ? fields[7] : "";
+
+    result[fields[6]] = std::move(socket_info);
+  }
 
-  return procDescriptors<decltype(descriptors)>(
-      process, L_procDescriptorsCallback, descriptors);
+  return Status(0);
 }
 
-Status procSocketInodeToFdMap(
-    const std::string& process,
-    std::unordered_map<std::string, std::string>& inode_to_fd_map) {
-  inode_to_fd_map.clear();
+Status procGetSocketInodeToSocketInfoMap(int family,
+                                         int protocol,
+                                         const std::string& pid,
+                                         SocketInodeToSocketInfoMap& result) {
+  std::string path = kLinuxProcPath + "/" + pid + "/net/";
+
+  switch (family) {
+  case AF_INET:
+    if (kLinuxProtocolNames.count(protocol) == 0) {
+      return Status(1,
+                    "Invalid family " + std::to_string(protocol) +
+                        " for AF_INET familiy");
+    } else {
+      path += kLinuxProtocolNames.at(protocol);
+    }
+    break;
+
+  case AF_INET6:
+    if (kLinuxProtocolNames.count(protocol) == 0) {
+      return Status(1,
+                    "Invalid protocol " + std::to_string(protocol) +
+                        " for AF_INET6 familiy");
+    } else {
+      path += kLinuxProtocolNames.at(protocol) + "6";
+    }
+    break;
+
+  case AF_UNIX:
+    if (protocol != IPPROTO_IP) {
+      return Status(1,
+                    "Invalid protocol " + std::to_string(protocol) +
+                        " for AF_UNIX familiy");
+    } else {
+      path += "unix";
+    }
+
+    break;
+
+  default:
+    return Status(1, "Invalid family " + std::to_string(family));
+  }
+
+  std::string content;
+  if (!osquery::readFile(path, content).ok()) {
+    return Status(1, "Could not open socket information from /proc");
+  }
 
-  auto L_procDescriptorsCallback =
-      [](const std::string& fd,
-         const std::string& link_name,
-         std::unordered_map<std::string, std::string>& inode_to_fd_map)
-      -> bool {
+  Status status(0);
+  switch (family) {
+  case AF_INET:
+  case AF_INET6:
+    status = procGetSocketInodeToSocketInfoMapInet(
+        family, protocol, path, content, result);
+    break;
+
+  case AF_UNIX:
+    status = procGetSocketInodeToSocketInfoMapUnix(path, content, result);
+    break;
+  }
+
+  return status;
+}
 
-    if (link_name.find("socket:[") != 0) {
+Status procGetSocketInodeToProcessInfoMap(const std::string& pid,
+                                          SocketInodeToProcessInfoMap& result) {
+  auto callback = [](const std::string& pid,
```

Comment:
This call back never returns false which seems weird to me. Is there no failure case of wanting to return false?

Reply:
The enumerate functions will stop executing the first time the callback returns false. Here if we find a file that isn't a socket that's okay and we just want to ignore it and continue the loop.